### ISSUE TITLE
URL Parameter Encoding

### ISSUE THREAD

**Issue by oursevendev** [BranchMetrics/iOS-Deferred-Deep-Linking-SDK#272]
Using marketing URL's, any parameters that are URL encoded get replaced by a "/". Any insight on maintaining url encoding when branch converts this into NSDictionary?

Thanks!

**Comment by oursevendev** [comment]
I have a temporary fix in place for our implementation needs using regex; however I feel there may be a more robust way to handle it; if not already one provided by your SDK.

Please & Thanks!

**Comment by derrickstaten** [comment]
@oursevendev could you provide an example? That will help us isolate where the issue is occurring and determine the best fix.

**Comment by oursevendev** [comment]
@derrickstaten sure, here's a more detailed explanation:
https://bnc.lt/a/key_live_{our_account_livekey}?feature=marketing&channel=email&agentKey=9941356&el=~~{urlEndodedToken}~~


"el=" (**encoded login** is the parameter that we have generated by our system and is url encoded): {
    "$one_time_use" = 0;
    "+click_timestamp" = 1453422181;
    "+clicked_branch_link" = 1;
    "+is_first_session" = 0;
    "+match_guaranteed" = 1;
    agentKey = 9941356;
    el = *{url encoded pieces get converted to a /, space etc. Any way to maintain the urlEncoded value? }*;
    "~channel" = email;
    "~feature" = marketing;
    "~id" = 212662481162655642;
}

  *The problem is that once it's on the branch side in iOS, as an NSDictionary; url encoded characters are removed and/or replaced!

  I have a work around in place, that is working for our current release of a product; however this would be a good to have fix to the branch SDK.

Cheers 🍻!

**Comment by ahmednawar** [comment]
@oursevendev thanks for letting us know. Can you send us a concrete string where this is happening -- the value that you are putting in the marketing link, what value you are getting, and what you expect to get passed in.

Also, would be great if you could share with us your workaround.

**Comment by ahmednawar** [comment]
@oursevendev we can't selectively decode some values. We will always decode any query parameters. You are always welcome to re-encode on the client.

**Comment by cmmouritsen** [comment]
I am running into an issue with this. For me, we are passing an email address as a parameter on a branch link. It's not very common, but when there's an email with a '+' character in it (i.e. `test+email@gmail.com`), then by the time Branch passes the parameters to my app it has become `test email@gmail.com`. I have also tried url encoding the email first (`test%2Bemail%40gmail.com`), but I got the same result. Any suggestions on how to address this?

**Comment by E-B-Smith** [comment]
Have you tried double URL encoding it?
That is, URL encode it, URL encode that, and then send it.  
When you get the value on the iOS side you will have to URL decode the value.

**Comment by E-B-Smith** [comment]
@cmmouritsen  See above ^^.

**Comment by cmmouritsen** [comment]
Thanks for the response. Something else may be going on. I keep getting the same params sent to my app (the first params that I used with the url) no matter what params I send. I've even tried another device and I still get the params from the first link that I clicked. It doesn't seem like that is how it is supposed to work. In any case, I'm no longer sure my second statement above is true about it not working when I encode the parameter. I realize this is a different issue, but any help would be appreciated.
